apps/sms/services/README.md

```diff
@@ -0,0 +1,91 @@
+# The layers in the Messages application
+
+## Terms explanation
+
+There are 3 layers whose interfaces managed by the Bridge (aka [threads.js](https://github.com/gaia-components/threads/)):
+
+* Services define high-level semantics to access the model. They are the only
+  interfaces that can be directly called by the views. Ideally one user action
+  has one matching service method.
+
+  Ideally a service doesn't call another service and calls only managers.
+  (but we can relax this rule in our case).
+
+  All together they form the Service Layer.
+
+* Managers define lower-level semantics. They'll be called by the Service layer.
+  Ideally their responsabilities are well-defined and they don't call each other.
+
+* Wrappers define interfaces to access mozAPI that are not available in workers.
+  Ideally they expose very similar APIs than their mozAPI counterparts because
+  their goal is to eventually disappear.
+
+## The Service Layer
+### ConversationService
+
+This service manages everything related to directly interacting with
+conversations and messages.
+
+Here are the operations needed:
+
+* Get All Conversations
+  - used in InboxView
+  - contains everything needed to display the Inbox (so including the "body", the "type", the "draft" and the "contact information".
+* Delete Conversations
+  -  used in InboxView
+* Delete Messages
+  -  used in ConversationView (and can in turn delete Conversations)
+* Mark Conversations As Read/Unread
+  -  Used in InboxView and ConversationView
+* Load 1 Conversation + its messages + contact(s) information(s)
+  -  Used in ConversationView
+* send an event when a message change and the local IDB (if existing) has been updated.
+* send an event when receiving a message and the local IDB (if existing) has been updated.
+* Load 1 Message
+  -  Used in ReportView and GroupView
+  - (If these views are really part of the ConversationPanel, maybe we don't need this operation)
+* find a conversation from number
+  - (if the redirection (new message -> conversation) is done from the service worker, this will be used from the service worker)
+
+### MessagingService
+
+This service will be used when a view does actions related to messaging functionality,
+especially sending messages.
+
+Here are the operations for this service:
+
+* send a message
+  - used from ConversationView and NewMessageView
+  - this service may use other services (eg: ConversationService to change model --
+    or underlying Manager ? or same JS file in both services that could do this ?)
+* the various events coming from mozMobileMessage
+* onSmsReceived (not an event, an actual method that will be called when we get
+  the system message from an hidden page or the SW)
+  - may use other services too
+
```

Comment:
> * MessagingService saves message in local DB using ConversationService (or ConversationManager, or directly the Threads js lib)
> * on success/failure, MessagingService updates message in local DB
MessagingService deletes current draft on success
> * MessagingService deletes current draft on success

Wondering if those MessagingService ---> ConversationService dependencies should be ConversationService ---> MessagingService instead so that ConversationService doesn't expose ```saveToDb\add``` methods directly? 

MessagingService fires events for such stages and ConversationService just listens for them. Exposing events doesn't seems to violate anything.

Regarding Steve's concern, as I understand we're choosing between two ways:

* The one from Julien
 * MozMobileMessageWrapperService is the plain wrapper around mozMobileMessage, here we should have as less business logic as we can. It sits inside some window context and exposes API via ```bridge.service.method```. It will be discontinued in the future, either because mozMobileMessage is available in worker or we'll get rid of it entirely.
 * MessagingService exposes high-level limited "gsm/messaging-only" subset of mozMobileMessage api + probably some pre/post processing; It internally calls MozMobileMessageWrapperService to perform API calls. It lives in SharedWorker;
 * ConversationService exposes high-level conversation/message/draft operations. Internally it uses MozMobileMessageWrapperService for some read-only mozMobileMessage operations and probably listen for high-level events from MessagingService (e.g. message-received event joined from MessagingAPI and system message, but it's a different story :))

* The one from Steve
 * MozMobileMessageWrapperService and MessagingService are joined into one MessagingService. Because it should access mozMobile API it _should live in window context_ so it can't be managed by Threads manager (the one from bridge.js) and can't use api available in worker (e.g. importScripts).
 * ConversationService is the same as in Julien's proposal but uses only MessagingService since we don't have separate low-level MozMobileMessageWrapperService.

So if I understood it correctly, for the longer term I'd prefer Julien's proposition since we can design MessagingService with worker env in mind and with more or less stable public interface. MessagingService can be shutdown in low-resource conditions. It can probably affect performance in some cases though, e.g. on-message-sending (MozMobileMessageWrapperService -> MessagingService -> ConversationService comparing to MessagingService -> ConversationService from Steve's proposal).

At the same time I'm ok with Steve's proposition if it's simpler/faster for us to implement as the first step and split them only when needed. Also it depends on the future of mozMobileMessage API, if the ultimate goal is to provide this API in worker environment, then it will be a bit simpler to migrate later.

So to summarize, I think we can go/live with any option unless some unforeseeable technical problems arise.